ml4gw/gw.py

```diff
@@ -285,6 +285,96 @@ def get_ifo_geometry(
     return torch.Tensor(tensors), torch.Tensor(vertices)
 
 
+def snr_frequency_series(
+    template: PSDTensor,
+    df: float,
+    psd: Optional[PSDTensor] = None,
+    strain: Optional[PSDTensor] = None,
+    highpass: Union[float, TensorType["frequency"], None] = None,
+    scale: float = 1e20,
+) -> PSDTensor:
+    """
+    Returns SNR as a function of frequency
+    """
+
+    if strain is None:
+        other = template
+    else:
+        other = strain
+
+    # TODO: handle scale issues by 1) dividing
+    # each tensor by the ASD separately and 2)
+    # multiplying by the sqrt of the scale factor.
+    # This is less optimal than our previous implementation
+    # because we're now doing 2 element-wise divides by
+    # the ASD rather than 1, so worth figuring out if
+    # there's a better, more consistent way to make the
+    # scale workout. All see TODO below
+    asd = psd**0.5
+    template = template * scale**0.5 / asd
+    other = other.conj() * scale**0.5 / asd
+    integrand = template * other
```

Comment:
Is this different than doing `integrand = template * other.conj() * scale / psd`? Do you run into precision issues?

Reply:
yeah that's exactly it, this is the most stable implementation I've found